<a href="https://colab.research.google.com/github/leejungp2/dl-study/blob/main/260531_05_pytorch_going_modular_cell_mode_video.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 05. Going Modular: Part 1 (cell mode)

* **노트북 개요**
  * 본 노트북은 [05. Going Modular](https://www.learnpytorch.io/05_pytorch_going_modular/) 섹션의 **Part 1/2**에 해당함
  * [notebook 04](https://www.learnpytorch.io/04_pytorch_custom_datasets/)의 요약본이며, 기존의 전통적인 Jupyter Notebook / Google Colab 방식으로 실행됨

* **참고: 두 파트의 구성 및 링크**
  1. [**05. Going Modular: Part 1 (cell mode)**](https://github.com/mrdbourke/pytorch-deep-learning/blob/main/going_modular/05_pytorch_going_modular_cell_mode.ipynb)
  2. [**05. Going Modular: Part 2 (script mode)**](https://github.com/mrdbourke/pytorch-deep-learning/blob/main/going_modular/05_pytorch_going_modular_script_mode.ipynb)
     * Part 1과 내용은 같으나, 주요 섹션들을 `data_setup.py`, `train.py`와 같은 독립된 **Python 스크립트 파일로 변환하는 기능이 추가**됨

* **두 파트로 나눈 이유 (Why two parts?)**
  * 무언가를 배우는 가장 좋은 방법은 간혹 그것이 다른 것과 **어떻게 다른지(differs)** 확인하는 것이기 때문임
  * 두 노트북을 한 화면에 나란히(side-by-side) 띄워놓고 실행해 보면 서로의 차이점을 명확히 알 수 있으며, 바로 그 차이점 속에 핵심 학습 내용이 담겨 있음

## 1. cell mode란 무엇인가? (What is cell mode?)
* **개념:** 이번 코스에서 지금까지 실행해 온 방식과 완전히 동일한 일반적인 노트북 형태를 의미함
* **특징:** 일부 셀(Cell)은 텍스트(마크다운)를 포함하고 있으며, 다른 셀들은 실행 가능한 소스코드를 포함하고 있음

## 2. 이 노트북(Part 1)과 script mode(Part 2)의 차이점은 무엇인가?
* **Part 1 (cell mode) 노트북의 역할**
  * [04. PyTorch Custom Datasets](https://www.learnpytorch.io/04_pytorch_custom_datasets/) 섹션에서 다룬 가장 유용한 핵심 코드들을 깔끔하게 정리하여 실행함
  * 처음부터 끝까지 정주행하면 04번 노트북에서 만들었던 이미지 분류 모델(TinyVGG)이 피자, 스테이크, 스시 이미지를 학습하는 과정을 그대로 재현할 수 있음
* **Part 2 (script mode)와의 핵심 차이점**
  * Part 2 노트북에는 각 섹션마다 셀 코드를 독립된 Python 스크립트 파일(`.py`)로 변환하기 위한 별도의 서브섹션(예: 2.1, 3.1, 4.1 등)이 추가되어 있다는 점이 가장 큰 차이임

## 3. 도움을 얻을 수 있는 곳 (Where can you get help?)
* **웹북 버전 가이드:** [learnpytorch.io - 05. PyTorch Going Modular](https://www.learnpytorch.io/05_pytorch_going_modular/) 섹션에서 책 형태로 제공되는 가이드를 확인 가능함
* **코스 전체 자료:** 이 코스의 다른 모든 학습 자료 및 소스코드는 [GitHub 리포지토리](https://github.com/mrdbourke/pytorch-deep-learning)에서 이용할 수 있음
* **트러블슈팅 및 질문:** 코드 실행 중 문제가 발생하면 코스 전용 [GitHub Discussions 페이지](https://github.com/mrdbourke/pytorch-deep-learning/discussions)에 질문을 남겨 해결할 수 있음
* **공식 커뮤니티:** PyTorch와 관련된 모든 의문점은 공식 [PyTorch 문서(Documentation)](https://pytorch.org/docs/stable/index.html) 및 집단지성이 모여있는 [PyTorch 개발자 포럼(Forums)](https://discuss.pytorch.org/)에서 매우 유용한 답변을 얻을 수 있음

## 0. Running a notebook in cell mode

As discussed, we're going to be running this notebook normally.

One cell at a time.

The code is from notebook 04, however, it has been condensed down to its core functionality.

## 1. Get data

We're going to start by downloading the same data we used in [notebook 04](https://www.learnpytorch.io/04_pytorch_custom_datasets/#1-get-data), the `pizza_steak_sushi` dataset with images of pizza, steak and sushi.

In [ ]:
# 데이터 다운로드 및 압축 해제 자동화 스크립트

import os
import zipfile

from pathlib import Path

import requests

# 데이터 저장 폴더의 기본 경로 설정 (data/pizza_steak_sushi)
data_path = Path("data/")
image_path = data_path / "pizza_steak_sushi"

# 데이터 폴더가 이미 존재하는지 확인 후, 없으면 새로 생성
if image_path.is_dir():
    print(f"{image_path} directory exists.") # 폴더가 이미 있으면 존재한다고 출력
else:
    print(f"Did not find {image_path} directory, creating one...")
    image_path.mkdir(parents=True, exist_ok=True) # 부모 폴더까지 포함하여 안전하게 디렉토리 생성

# GitHub에서 피자, 스테이크, 스시 압축 데이터(.zip) 다운로드
with open(data_path / "pizza_steak_sushi.zip", "wb") as f:
    # requests 라이브러리로 URL에 있는 바이너리 압축 파일 데이터를 가져옴
    request = requests.get("https://github.com/mrdbourke/pytorch-deep-learning/raw/main/data/pizza_steak_sushi.zip")
    print("Downloading pizza, steak, sushi data...")
    f.write(request.content) # 다운로드한 파일 내용을 바이너리 모드로 로컬에 기록

# 다운로드 완료된 zip 파일 압축 해제
with zipfile.ZipFile(data_path / "pizza_steak_sushi.zip", "r") as zip_ref:
    print("Unzipping pizza, steak, sushi data...")
    zip_ref.extractall(image_path) # 지정해 둔 이미지 폴더(image_path) 내부에 압축 해제

# 디스크 용량 관리를 위해 사용이 끝난 원본 zip 파일 삭제
os.remove(data_path / "pizza_steak_sushi.zip")

data/pizza_steak_sushi directory exists.
Unzipping pizza, steak, sushi data...


<추가 설명>

① 파일 경로 관리에 pathlib.Path를 쓰는 이유
기존의 os.path.join() 방식은 운영체제(Windows의 \, Linux/Mac의 /)마다 슬래시 방향이 달라 코드가 꼬이기 쉽음

Path("data/") / "pizza_steak_sushi"처럼 나누기 기호(/)를 사용해 직관적으로 경로를 결합할 수 있으며, 어떤 OS에서 실행하더라도 시스템에 맞게 올바른 경로 파일 표현식으로 변환됨

mkdir(parents=True, exist_ok=True)에서 parents=True는 상위 폴더가 없으면 상위 폴더까지 연쇄적으로 만들어주고, exist_ok=True는 폴더가 이미 있어도 에러를 터뜨리지 않고 조용히 넘어가게 해줌

② 안전한 파일 처리를 위한 with open 문 (Context Manager)
파일을 열고 작업을 마친 뒤 f.close()를 직접 호출하지 않으면 시스템 메모리 누수나 파일 잠김 현상이 발생할 수 있음

with 블록을 사용하면 내부에서 오류(Exception)가 발생하더라도 블록을 벗어나는 순간 자동으로 파일을 안전하게 닫아(Close)주므로 필수적인 코딩 패턴임

압축 파일 같은 바이너리 형태의 데이터를 다운로드하여 저장할 때는 일반 텍스트 모드(w)가 아닌 바이너리 쓰기 모드(wb)를 반드시 명시해야 데이터가 깨지지 않음

③ 파이프라인 자동화 관점에서의 의의
이 코드는 모델 학습을 시작하기 전, 로컬 환경이나 Google Colab 세션이 초기화되었을 때 데이터셋 준비 작업을 코드로 완전 자동화해 줌

코드의 마지막 줄에서 os.remove()를 통해 압축 파일을 즉시 지워주는 것은 한정된 클라우드 인스턴스나 로컬 디스크의 저장 공간(Storage) 낭비를 최소화하는 좋은 습관임

In [ ]:
# Setup train and testing paths
train_dir = image_path / "train"
test_dir = image_path / "test"

train_dir, test_dir

(PosixPath('data/pizza_steak_sushi/train'),
 PosixPath('data/pizza_steak_sushi/test'))

## 2. 데이터셋 및 데이터로더 생성

이제 이미지 데이터셋을 PyTorch의 Dataset 및 DataLoader로 변환해 보겠습니다

In [ ]:
# ImageFolder를 활용한 기본 데이터셋 생성 스크립트
from torchvision import datasets, transforms

# 이미지를 64x64 크기로 조절하고 텐서로 변환하는 기본 전처리 파이프라인 생성
data_transform = transforms.Compose([
    transforms.Resize((64, 64)), # 모든 이미지의 해상도를 가로 64, 세로 64 픽셀로 통일
    transforms.ToTensor(),       # PIL 이미지나 NumPy 배열을 PyTorch 텐서로 변환 (0~1 스케일링 포함)
])

# ImageFolder를 사용하여 훈련 데이터셋 생성
train_data = datasets.ImageFolder(root=train_dir, # 훈련용 이미지들이 정렬되어 있는 대상 폴더 경로
                                  transform=data_transform, # 이미지에 적용할 전처리 설정
                                  target_transform=None) # 정답 레이블(Label)에 추가 변형을 주지 않음 (필요시 사용)

# ImageFolder를 사용하여 테스트 데이터셋 생성
test_data = datasets.ImageFolder(root=test_dir, # 테스트용 이미지들이 정렬되어 있는 대상 폴더 경로
                                 transform=data_transform) # 훈련셋과 동일한 기본 전처리 적용

# 생성된 훈련 및 테스트 데이터셋의 요약 정보 출력
print(f"Train data:\n{train_data}\nTest data:\n{test_data}")

Train data:
Dataset ImageFolder
    Number of datapoints: 225
    Root location: data/pizza_steak_sushi/train
    StandardTransform
Transform: Compose(
               Resize(size=(64, 64), interpolation=bilinear, max_size=None, antialias=True)
               ToTensor()
           )
Test data:
Dataset ImageFolder
    Number of datapoints: 75
    Root location: data/pizza_steak_sushi/test
    StandardTransform
Transform: Compose(
               Resize(size=(64, 64), interpolation=bilinear, max_size=None, antialias=True)
               ToTensor()
           )


<결과 해석>

① 데이터 포인트(Datapoints) 구성비 확인

* 훈련 데이터 225장, 테스트 데이터 75장으로 총 300장의 이미지가 로드되었음
* 훈련셋과 테스트셋의 비율이 정확히 **75% : 25% ($3:1$)**로 분할되어 있어, 소규모 데이터셋 환경에서 모델을 학습시키고 평가하기에 표준적이고 이상적인 비율을 유지하고 있음

② 변환 인터폴레이션(`interpolation=bilinear`)의 의미

* 원본 이미지의 크기를 $64 \times 64$로 줄일 때, 픽셀과 픽셀 사이의 공백을 주변 4개 픽셀의 쌍선형 가중치 평균값으로 채워 넣는 **쌍선형 보간법(Bilinear Interpolation)**이 적용됨
* 이미지 축소 시 계단 현상(깨짐)을 줄이고 부드럽게 크기를 변경해 주는 가장 대중적이고 연산 효율이 좋은 기본 알고리즘임

③ 안티앨리어싱(`antialias=True`) 활성화 상태

* 이미지를 대폭 축소할 때 고주파 신호가 왜곡되어 노이즈가 발생하는 앨리어싱(Aliasing) 현상을 방지하는 필터가 켜져 있음
* 이 옵션 덕분에 이미지가 $64 \times 64$로 작아져도 사물의 뼈대나 주요 특징 선이 뭉개지지 않고 깔끔하게 보존되어 모델이 사물의 특징을 더 쉽게 학습할 수 있도록 도움을 줌

In [ ]:
# 1. 클래스 이름 및 매핑 딕셔너리 확인
class_names = train_data.classes      # 데이터셋의 클래스 이름들을 리스트 형태로 가져옴 (예: ['pizza', 'steak', 'sushi'])
class_names


['pizza', 'steak', 'sushi']

In [ ]:
# Can also get class names as a dict
class_dict = train_data.class_to_idx  # 각 클래스 이름이 어떤 정수 인덱스와 매핑되어 있는지 딕셔너리로 확인
class_dict

{'pizza': 0, 'steak': 1, 'sushi': 2}

In [ ]:
# 2. 데이터셋의 총 샘플 수 확인
len(train_data), len(test_data)       # 훈련 데이터셋과 테스트 데이터셋의 데이터 개수(길이)를 튜플 형태로 확인

(225, 75)

In [ ]:
# 3. PyTorch Dataset을 DataLoader로 변환 (미니배치 생성 및 멀티프로세싱 설정)
from torch.utils.data import DataLoader

train_dataloader = DataLoader(dataset=train_data,
                              batch_size=1,    # 한 번의 연산(배치)에 몇 개의 데이터를 사용할지 설정 (현재는 1개)
                              num_workers=1,   # 데이터 로딩에 사용할 하위 프로세스(CPU 코어) 개수
                              shuffle=True)    # 매 에폭마다 훈련 데이터의 순서를 무작위로 섞어 모델의 과적합을 방지

test_dataloader = DataLoader(dataset=test_data,
                             batch_size=1,
                             num_workers=1,
                             shuffle=False)   # 테스트 데이터는 평가용이므로 무작위로 섞을 필요가 없음

train_dataloader, test_dataloader

(<torch.utils.data.dataloader.DataLoader at 0x791aba77ade0>,
 <torch.utils.data.dataloader.DataLoader at 0x791becb44170>)

In [ ]:
# 4. DataLoader에서 단일 미니배치 샘플 추출 및 차원 확인
img, label = next(iter(train_dataloader)) # DataLoader를 이터레이터로 변환 후 첫 번째 배치(샘플 1개)를 꺼냄

# 꺼내온 이미지와 레이블의 텐서 형태(Shape) 출력
print(f"Image shape: {img.shape} -> [batch_size, color_channels, height, width]")
print(f"Label shape: {label.shape}")

💡 추가 상세 설명

① 텐서 차원 분석: `[batch_size, color_channels, height, width]`

* **`img.shape` 결과 예측:** `batch_size=1`로 설정했기 때문에 출력 결과는 `torch.Size([1, 3, 64, 64])`가 됩니다.
* **차원별 의미:**
  * **`1` (Batch Size):** 한 번의 순방향 연산(Forward Pass)에 동시에 투입되는 이미지의 개수입니다.
  * **`3` (Color Channels):** RGB 컬러 채널의 수입니다.
  * **`64, 64` (Height, Width):** 앞서 전처리(`data_transform`)에서 지정한 이미지의 가로 및 세로 해상도입니다.

② `DataLoader` 파라미터 튜닝 포인트

* **`batch_size=1`의 한계:** 배치의 크기가 1이면 모델이 이미지 1장을 볼 때마다 가중치를 업데이트하므로, 전체적인 연산 속도가 극도로 떨어지고 경사하강법의 경로가 심하게 요동칩니다. 하드웨어(GPU) 메모리가 허용하는 한 일반적으로 **32, 64, 128** 등의 2의 거듭제곱 단위로 늘려서 연산 병렬성을 높이는 것이 좋습니다.
* **`num_workers` 설정:** CPU가 배경에서 데이터를 미리 로드(Prefetch)하여 GPU가 쉬지 않고 연산할 수 있도록 돕는 파라미터입니다. 주피터 노트북 환경이나 로컬 환경의 특성에 따라 너무 높으면 에러가 날 수 있으므로, 보통 시스템의 CPU 물리 코어 수에 맞춰 적절히 지정해 주어야 병목 현상이 생기지 않습니다.

③ 데이터 추출 코드 흐름 이해 (`next(iter(...))`)

* `train_dataloader` 자체는 파이썬에서 대용량 데이터를 효율적으로 관리하는 파이프라인 구조이기 때문에, `dataloader[0]`처럼 인덱싱으로 데이터를 꺼낼 수 없습니다.
* `iter(train_dataloader)`를 통해 데이터를 차례대로 공급할 수 있는 **반복자(Iterator)** 형태로 변환한 뒤, `next()` 함수를 호출함으로써 맨 첫 번째 미니배치 데이터(이미지 뭉치와 레이블 뭉치)를 가로채 구조를 검사하는 방식입니다.

## 2.1 데이터셋 및 데이터로더 생성 (스크립트 모드)

> 175. 우리 파이토치 데이터로더를 생성하기 위한 파이썬 스크립트 만들기

데이터로더를 생성하는 `.py` 파일을 만들기 위해 주피터 매직 함수(Jupyter magic function)를 사용해 봅시다.

주피터 매직 명령어인 `%%writefile 파일이름`을 사용하면 코드 셀의 내용들을 파일로 저장할 수 있습니다.
* 참고 링크: https://ipython.readthedocs.io/en/stable/interactive/magics.html#cellmagic-writefile

In [ ]:
# Create a directory going_modular scripts
import os
os.makedirs("going_modular")

In [ ]:
# 데이터 로더 생성 스크립트 모듈화 코드 해석

%%writefile going_modular/data_setup.py
"""
이미지 분류 데이터를 위한 PyTorch DataLoader 생성 기능을 포함하는 모듈입니다.
"""
import os

from torchvision import datasets, transforms
from torch.utils.data import DataLoader

# 현재 시스템 환경의 최대 CPU 코어 수를 자동으로 감지하여 기본 워커 수로 설정
NUM_WORKERS = os.cpu_count()

def create_dataloaders(
  train_dir: str,
  test_dir: str,
  transform: transforms.Compose,
  batch_size: int,
  num_workers: int=NUM_WORKERS  # 값을 전달받지 않으면 감지된 CPU 코어 수를 기본값으로 사용
):
  """훈련 및 테스트용 DataLoader를 생성하는 함수입니다.

  훈련 및 테스트 디렉토리 경로를 입력받아 PyTorch Dataset으로 변환한 뒤,
  최종적으로 PyTorch DataLoader로 만들어 반환합니다.

  Args (입력 매개변수):
    train_dir: 훈련 데이터 디렉토리 경로 (문자열)
    test_dir: 테스트 데이터 디렉토리 경로 (문자열)
    transform: 훈련 및 테스트 데이터에 적용할 torchvision 전처리 파이프라인
    batch_size: 각 DataLoader에서 처리할 배치 크기 (미니배치 당 샘플 수)
    num_workers: 각 DataLoader에서 데이터 로딩에 사용할 서브프로세스 개수

  Returns (반환값):
    (train_dataloader, test_dataloader, class_names) 형태의 튜플.
    여기서 class_names는 타겟 클래스들의 이름을 담은 리스트입니다.

    사용 예시:
      train_dataloader, test_dataloader, class_names = create_dataloaders(
        train_dir="path/to/train_dir",
        test_dir="path/to/test_dir",
        transform=some_transform,
        batch_size=32,
        num_workers=4
      )
  """
  # 1. ImageFolder를 사용해 각각 훈련 및 테스트 데이터셋(Dataset) 객체 생성
  train_data = datasets.ImageFolder(train_dir, transform=transform)
  test_data = datasets.ImageFolder(test_dir, transform=transform)

  # 2. 데이터셋 폴더 구조로부터 자동으로 타겟 클래스 이름 리스트 추출
  class_names = train_data.classes

  # 3. 데이터셋을 미니배치 형태의 훈련용 DataLoader로 변환
  train_dataloader = DataLoader(
      train_data,
      batch_size=batch_size,
      shuffle=True,          # 학습 데이터는 매 에폭마다 무작위로 섞음
      num_workers=num_workers,
      pin_memory=True        # GPU 연산 속도 향상을 위해 고정 메모리 영역(Page-locked) 활용
                             # 자세한 내용은 공식 문서 참고: [https://pytorch.org/docs/stable/data.html](https://pytorch.org/docs/stable/data.html)
  )

  # 4. 데이터셋을 미니배치 형태의 테스트용 DataLoader로 변환
  test_dataloader = DataLoader(
      test_data,
      batch_size=batch_size,
      shuffle=False,         # 평가는 순서대로 진행하므로 섞지 않음
      num_workers=num_workers,
      pin_memory=True
  )

  # 5. 생성된 요소들을 튜플 형태로 최종 반환
  return train_dataloader, test_dataloader, class_names

Writing going_modular/data_setup.py


💡 추가 상세 설명

① os.cpu_count()의 자동화 관점 이점
: 코드를 작성한 컴퓨터와 실제로 코드를 돌리는 서버(Colab, AWS 등)의 CPU 환경은 제각각 다릅니다.

os.cpu_count()를 기본값으로 지정해 두면 디바이스 환경이 바뀌더라도 수동으로 코드를 수정할 필요 없이, 시스템 하드웨어 자원을 자동으로 100% 끌어다 쓸 수 있어 이식성이 뛰어난 독립 모듈이 완성됩니다.

② pin_memory=True 옵션의 역할과 멀티프로세싱 최적화

pin_memory=True로 설정하면 CPU 메모리 영역에서 호스트의 페이지 고정 메모리(Page-locked Memory)를 확보합니다.

이 영역에 적재된 이미지 데이터는 가상 메모리로 빠지지 않고 바로 컴퓨터의 주 메모리에 고정되므로, GPU(CUDA 디바이스)로 데이터를 전송(Transfer)할 때 전송 속도가 훨씬 빨라집니다. 특히 GPU 연산 능력이 데이터 로딩 속도보다 빠를 때 병목을 해결하는 필수 옵션입니다.

③ 모듈화 코딩의 완성: 독스트링(Docstring)과 타입 힌팅(Type Hinting)

파라미터 옆에 적힌 : str, : int와 같은 문법은 변수의 데이터 타입을 명시하는 타입 힌트(Type Hint)입니다.

아래에 삼중 따옴표(""")로 작성된 독스트링(Docstring)은 다른 개발자나 미래의 내가 이 함수를 재사용할 때 함수 내부에 들어가지 않고도 어떤 인자를 전달해야 하고, 결과값으로 무엇이 나오는지 명확히 파악할 수 있도록 돕는 프로덕션 수준의 표준 설계 방식입니다.

In [ ]:
# 생성한 모듈(data_setup)을 활용한 DataLoader 로드 스크립트

# 1. 앞서 %%writefile로 분리 저장한 going_modular 폴더 내의 data_setup.py 모듈을 임포트
from going_modular import data_setup

# 2. 모듈 내부의 create_dataloaders 함수를 호출하여 데이터로더 및 클래스 이름 가져오기
train_dataloader, test_dataloader, class_names = data_setup.create_dataloaders(
    train_dir=train_dir,      # 훈련 데이터 경로
    test_dir=test_dir,        # 테스트 데이터 경로
    transform=data_transform,  # 이미지 전처리 파이프라인 (Resize, ToTensor 등)
    batch_size=32             # 미니배치 크기를 32로 설정 (기존 1에서 32로 확장)
)

# 3. 생성된 객체들의 정보 확인 및 출력
train_dataloader, test_dataloader, class_names

(<torch.utils.data.dataloader.DataLoader at 0x791aba80a570>,
 ['pizza', 'steak', 'sushi'])

## 3. 모델 만들기 (TinyVGG)

> 176. 모델 구축 코드를 파이썬 스크립트로 변환하기

우리는 노트북 04에서 사용했던 것과 동일한 모델인 CNN Explainer 웹사이트의 TinyVGG를 사용할 것입니다.

노트북 04와 비교하여 여기서 변경된 유일한 점은 [Google의 파이썬 스타일 가이드(Google's Style Guide for Python)](https://google.github.io/styleguide/pyguide.html#384-classes)를 준수하는 독스트링(docstring)이 추가되었다는 것입니다.

In [ ]:
# TinyVGG 모델 신경망 구조 코드 해석

import torch

from torch import nn

class TinyVGG(nn.Module):
  """TinyVGG 신경망 구조를 생성합니다.

  CNN Explainer 웹사이트에 소개된 TinyVGG 구조를 PyTorch로 재현합니다.
  원래 구조는 다음 링크에서 확인할 수 있습니다: https://poloclub.github.io/cnn-explainer/

  Args (매개변수):
    input_shape: 입력 이미지의 채널 수를 나타내는 정수 (예: RGB 이미지는 3).
    hidden_units: 레이어 사이의 은닉 뉴런(채널) 수를 나타내는 정수.
    output_shape: 최종 출력 뉴런 수를 나타내는 정수 (분류할 클래스 개수).
  """
  def __init__(self, input_shape: int, hidden_units: int, output_shape: int) -> None:
      super().__init__()

      # 1번째 합성곱 블록 (Convolutional Block 1)
      self.conv_block_1 = nn.Sequential(
          nn.Conv2d(in_channels=input_shape,   # 입력 채널 수
                    out_channels=hidden_units, # 출력 채널 (은닉 뉴런) 수
                    kernel_size=3,             # 이미지를 훑는 필터(커널)의 크기 (3x3 화소 크기 격자)
                    stride=1,                  # 필터가 한 번에 이동하는 한 칸씩의 보폭 (기본값)
                    padding=0),                # 패딩 없음. "valid"(패딩 없음) 또는 "same"(입출력 크기 동일) 혹은 특정 정수 지정 가능
          nn.ReLU(),                           # 비선형 활성화 함수 활성화
          nn.Conv2d(in_channels=hidden_units,
                    out_channels=hidden_units,
                    kernel_size=3,
                    stride=1,
                    padding=0),
          nn.ReLU(),
          nn.MaxPool2d(kernel_size=2,
                        stride=2) # 가로세로 크기를 절반으로 줄이는 맥스 풀링 (기본 보폭 stride 값은 커널 크기와 동일하게 2로 설정됨)
      )

      # 2번째 합성곱 블록 (Convolutional Block 2)
      self.conv_block_2 = nn.Sequential(
          nn.Conv2d(hidden_units, hidden_units, kernel_size=3, padding=0), # 1블록에서 확장된 은닉 채널 크기를 그대로 유지하며 연산
          nn.ReLU(),
          nn.Conv2d(hidden_units, hidden_units, kernel_size=3, padding=0),
          nn.ReLU(),
          nn.MaxPool2d(2) # 크기를 한 번 더 절반으로 줄여 특징을 압축
      )

      # 최종 분류기 레이어 (Classifier Stage)
      self.classifier = nn.Sequential(
          nn.Flatten(), # 다차원 텐서(배치, 채널, 높이, 너비)를 완전연결층에 넣기 위해 1차원 벡터로 평탄화

          # 이 in_features의 크기(hidden_units * 13 * 13)는 어떻게 도출되었는가?
          # 신경망의 각 레이어를 거칠 때마다 입력 데이터의 크기(H, W)가 압축되고 변형되기 때문임.
          # (64x64 이미지가 두 번의 Conv와 한 번의 MaxPool을 거치며 특징 지도가 13x13 픽셀 크기로 줄어듦)
          nn.Linear(in_features=hidden_units*13*13, # 평탄화된 전체 입력 특징 수
                    out_features=output_shape)      # 최종 예측할 클래스 개수(Logits 값 출력)
      )

  def forward(self, x: torch.Tensor):
      # 순방향 연산 수행: 입력 데이터 x를 각 블록에 순서대로 통과시킴
      x = self.conv_block_1(x)
      x = self.conv_block_2(x)
      x = self.classifier(x)
      return x

💡 추가 상세 설명


① `in_features=hidden_units * 13 * 13` 차원 계산 원리

$64 \times 64$ 크기의 이미지가 각 레이어를 통과할 때 패딩이 0(`padding=0`)이고 커널 크기가 3(`kernel_size=3`)이면, 가로세로 크기가 각각 2픽셀씩 줄어든다.

- **`conv_block_1` 통과 과정:**
   * 입력: $64 \times 64$
   * 첫 번째 Conv2d: $64 - 3 + 1 = 62 \times 62$
   * 두 번째 Conv2d: $62 - 3 + 1 = 60 \times 60$
   * MaxPool2d (크기 2, 보폭 2): 절반으로 압축되어 $30 \times 30$이 됨
- **`conv_block_2` 통과 과정:**
   * 첫 번째 Conv2d: $30 - 3 + 1 = 28 \times 28$
   * 두 번째 Conv2d: $28 - 3 + 1 = 26 \times 26$
   * MaxPool2d (크기 2, 보폭 2): 절반으로 압축되어 최종 **$13 \times 13$** 수치가 도출됨
- **`Flatten` 이후:**
   * 최종 채널 개수인 `hidden_units`에 남은 이미지 공간 크기 $13 \times 13$을 모두 곱한 값이 1차원 결합 벡터의 크기가 되므로 `hidden_units * 13 * 13` 구조를 선언해야 에러가 나지 않는다.

② 연산자 퓨전 (Operator Fusion) 주석의 의미

* 주석 처리된 맨 아랫줄 `return self.classifier(self.block_2(self.block_1(x)))` 코드는 가독성은 조금 떨어질 수 있으나 변수 재할당 연산을 줄여준다.
* PyTorch 내부 컴파일러나 하드웨어 단에서 중간 텐서 값을 메모리에 썼다 읽었다 하는 오버헤드를 줄이고, 연속된 연산을 하나의 커널로 묶어 처리(Fusion)할 수 있는 최적화 여지를 열어주는 코딩 스타일이다.

In [ ]:
# 모델 인스턴스 생성 및 디바이스 할당 스크립트 코드 해석

import torch

# 1. 연산을 수행할 하드웨어 장치(Device) 설정
# GPU(CUDA)가 사용 가능하면 "cuda"를 선택하고, 사용할 수 없다면 기본 CPU 환경으로 설정함
device = "cuda" if torch.cuda.is_available() else "cpu"

# 2. 모델 인스턴스 생성 (객체화)
torch.manual_seed(42) # 가중치(Weights) 초기화 시 결과의 재현성을 보장하기 위해 무작위 난수 시드 고정

model_0 = TinyVGG(input_shape=3, # 입력 이미지의 컬러 채널 수 (RGB 이미지이므로 3)
                  hidden_units=10, # 컨볼루션 레이어 사이의 은닉 뉴런(채널) 수를 10으로 설정
                  output_shape=len(train_data.classes)).to(device) # 출력 뉴런 수는 학습 데이터의 클래스 개수(3개: pizza, steak, sushi)로 설정하고, 모델의 가중치를 앞서 설정한 연산 장치(GPU 또는 CPU)의 메모리로 전송함

# 3. 생성된 모델 구조 확인 및 출력
model_0

TinyVGG(
  (conv_block_1): Sequential(
    (0): Conv2d(3, 10, kernel_size=(3, 3), stride=(1, 1))
    (1): ReLU()
    (2): Conv2d(10, 10, kernel_size=(3, 3), stride=(1, 1))
    (3): ReLU()
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (conv_block_2): Sequential(
    (0): Conv2d(10, 10, kernel_size=(3, 3), stride=(1, 1))
    (1): ReLU()
    (2): Conv2d(10, 10, kernel_size=(3, 3), stride=(1, 1))
    (3): ReLU()
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (classifier): Sequential(
    (0): Flatten(start_dim=1, end_dim=-1)
    (1): Linear(in_features=1690, out_features=3, bias=True)
  )
)

💡 추가 상세 설명

### ① 디바이스 동적 할당 (`.to(device)`)의 중요성
* 딥러닝에서 대규모 텐서 연산을 빠르게 처리하려면 연산의 무대를 CPU가 아닌 GPU 메모리로 옮겨야 합니다.
* 하드웨어 환경이 매번 바뀔 수 있기 때문에, 삼항 연산자를 이용해 코드가 구동되는 머신의 GPU 환경(`cuda.is_available()`)을 실시간으로 확인한 뒤 자동으로 타겟 메모리에 모델 가중치를 바인딩해 주는 기법은 필수적인 개발 패턴입니다.

### ② `torch.manual_seed(42)`를 작성하는 이유
* 인공지능 신경망은 맨 처음 학습을 시작할 때 내부 필터(Kernel)와 가중치 값을 무작위 난수(Random Number)로 생성합니다.
* 시드(Seed)를 고정하지 않으면 코드를 돌릴 때마다 매번 출발점(초기 가중치 상태)이 달라져, 모델의 성능이 개선된 것인지 우연히 초기화가 잘 된 것인지 객관적으로 평가하기 어려워집니다. 실험 환경을 통제하고 항상 일관된 실험 결과(Reproducibility)를 얻기 위해 시드를 고정합니다.

### ③ 출력 뉴런 크기의 동적 결정 (`len(train_data.classes)`)
* 출력 레이어의 차원을 숫자 `3`으로 하드코딩(Hard-coding)하는 대신, 데이터셋의 속성 배열 길이(`len(...)`)를 참조하도록 설계했습니다.
* 이렇게 구현하면 나중에 데이터셋에 새로운 음식 카테고리(예: 디저트, 버거 등)가 추가되거나 클래스가 줄어들더라도, 모델 생성 부의 코드를 전혀 건드리지 않고 유연하게 유지보수할 수 있습니다.

💡 TinyVGG 구조 출력 결과(Architecture Output) 상세 분석

* **전체 네트워크 구성**: 2개의 합성곱 블록(`conv_block_1`, `conv_block_2`)과 1개의 최종 분류기 레이어(`classifier`)가 순차적으로 연결된 구조

---

컨볼루션 블록 1 (`conv_block_1`)

* **`(0): Conv2d(3, 10, kernel_size=(3, 3), stride=(1, 1))`**
  * 입력 채널: 3 (RGB 컬러 이미지)
  * 출력 채널: 10 (10개의 가중치 필터로 10차원의 특징 맵 추출)
  * 커널 크기: 3x3 격자 크기로 이미지를 훑음
  * 보폭(Stride): 1칸씩 이동하며 조밀하게 연산
* **`(1) & (3): ReLU()`**
  * 비선형 활성화 함수를 도입하여 모델의 표현력을 높임
* **`(2): Conv2d(10, 10, ...)`**
  * 앞서 추출된 10차원 특징을 그대로 이어받아 사물의 기하학적 패턴을 한 번 더 누적
* **`(4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)`**
  * 2x2 영역 내에서 가장 큰 값만 남기는 맥스 풀링 연산
  * 이미지의 가로세로 크기가 정확히 절반으로 압축됨
  * `ceil_mode=False`: 크기 계산 시 소수점이 발생하면 내림(Floor) 처리하여 자투리 픽셀을 버림

---

컨볼루션 블록 2 (`conv_block_2`)

* **공간 정보의 지속적 압축**: 1번 블록을 거치며 해상도가 대폭 줄어든 텐서를 입력받음
* **채널 크기 유지**: 입력과 출력 채널이 모두 `10`으로 고정됨 (`Conv2d(10, 10)`)
  * 연산 파라미터의 급증을 막으면서, 모델의 수용 영역(Receptive Field)을 넓혀 고차원적 특징 파악

---

최종 분류기 레이어 (`classifier`)

* **`(0): Flatten(start_dim=1, end_dim=-1)`**
  * 배치 차원(0번째 차원)은 그대로 유지함
  * 데이터 입력 표준에 맞추기 위해 이미지 채널과 공간 정보(H, W)가 섞인 다차원 텐서를 1차원 벡터로 일렬 평탄화
* **`(1): Linear(in_features=1690, out_features=3, bias=True)`**
  * `in_features=1690`: 컨볼루션 연산을 모두 거치고 남은 최종 텐서의 크기 ($10\text{채널} \times 13\text{높이} \times 13\text{너비} = 1690$)
  * `out_features=3`: 최종 도출할 예측 클래스 개수 (피자, 스테이크, 스시 총 3종류)
  * `bias=True`: 선형 회귀 연산 시 y절편에 해당하는 편향(Bias) 가중치를 학습 대상에 포함

To test our model let's do a single forward pass (pass a sample batch from the training set through our model).

In [ ]:
# 1. DataLoader에서 이미지와 레이블 한 배치(Batch) 가져오기
# 반복자(Iterator)를 생성한 뒤 next()를 호출하여 첫 번째 미니배치 데이터([32, 3, 64, 64])를 추출함
img_batch, label_batch = next(iter(train_dataloader))

# 2. 배치에서 단일 이미지 추출 및 모델 입력 규격에 맞게 차원 확장(Unsqueeze)
# 배치 내 첫 번째 이미지([3, 64, 64])를 선택한 후, 0번째 차원에 1을 추가하여 [1, 3, 64, 64] 형태로 만듦 (레이블은 정수형 스칼라로 유지)
img_single, label_single = img_batch[0].unsqueeze(dim=0), label_batch[0]
print(f"Single image shape: {img_single.shape}\n")

# 3. 단일 이미지에 대한 순방향 연산(Forward Pass) 및 추론 진행
model_0.eval() # 모델을 평가(추론) 모드로 전환 (드롭아웃, 배치 정규화 등의 레이어가 가동 중지됨)
with torch.inference_mode(): # 기울기(Gradient) 계산을 비활성화하여 메모리를 절약하고 연산 속도를 최적화함
    pred = model_0(img_single.to(device)) # 이미지를 연산 장치(GPU/CPU)로 전송한 후 모델에 입력하여 가중치 연산 수행

# 4. 연산 과정 출력 및 모델의 출력값(Logits)을 확률(Probs)과 최종 예측 레이블(Label)로 변환
print(f"Output logits:\n{pred}\n") # 모델이 가공 없이 내놓은 원시 예측값(Logits) 출력

# Softmax 함수를 적용하여 3개 클래스의 합이 1(100%)이 되도록 확률 분포값으로 변환
print(f"Output prediction probabilities:\n{torch.softmax(pred, dim=1)}\n")

# argmax 함수를 사용해 확률 분포 중 가장 높은 확률을 가진 인덱스(가장 신뢰도가 높은 클래스 번호)를 추출
print(f"Output prediction label:\n{torch.argmax(torch.softmax(pred, dim=1), dim=1)}\n")
print(f"Actual label:\n{label_single}") # 데이터셋에 기록되어 있던 실제 정답 레이블(Ground Truth) 출력

Single image shape: torch.Size([1, 3, 64, 64])

Output logits:
None



TypeError: softmax() received an invalid combination of arguments - got (NoneType, dim=int), but expected one of:
 * (Tensor input, int dim, torch.dtype dtype = None, *, Tensor out = None)
 * (Tensor input, name dim, *, torch.dtype dtype = None)


💡 추가 상세 설명

unsqueeze(dim=0)의 차원 맞춤 역할
- PyTorch의 모든 2D CNN 모델은 단일 이미지가 아닌 [배치_크기, 채널, 높이, 너비] 형태의 4차원 뭉치(Batch) 데이터를 입력받도록 설계되어 있습니다.
- 배치에서 이미지 1장만 슬라이싱(img_batch[0])하면 [3, 64, 64]인 3차원 텐서가 되어 모델이 입력을 거부하고 에러를 발생시킵니다
- .unsqueeze(dim=0)를 통해 0번째 자리에 가짜 배치 차원을 인위적으로 삽입하여 [1, 3, 64, 64] 형태로 규격을 강제 매칭해 주는 필수 전처리 작업입니다.

model.eval()과 torch.inference_mode()의 최적화 조합
- model.eval(): 모델을 평가(추론) 모드로 내부 설정을 전환합니다. 학습 시에만 동작해야 하는 드롭아웃(Dropout)이나 배치 정규화(Batch Normalization) 등의 레이어가 추론 시 고정된 결정론적 방식으로 작동하도록 제어합니다.
- torch.inference_mode(): 과거 버전의 torch.no_grad()를 대체하는 더 빠르고 최적화된 환경입니다. 역전파에 필요한 기울기(Gradient) 계산 및 내부 텐서 추적 정보를 완전히 차단하여 메모리 오버헤드를 막고 추론 속도를 극대화합니다.

출력 데이터의 3단계 파이프라인 흐름 (Logits $\rightarrow$ Probs $\rightarrow$ Label)
- 로짓 (Logits): 모델의 최종 선형 결합층이 출력한 가공되지 않은 상태의 원시 점수 벡터입니다. 범위가 $-\infty$에서 $+\infty$ 사이이므로 직관적인 확률로 해석할 수 없습니다.
- 예측 확률 (Prediction Probabilities): 로짓을 torch.softmax(dim=1) 함수에 통과시키면 모든 클래스의 점수 합이 정확히 1(100%)이 되는 $0 \sim 1$ 사이의 값으로 스케일링되어 변환됩니다. 모델이 각 음식 카테고리에 대해 느끼는 신뢰도 점수로 활용됩니다.
- 예측 레이블 (Prediction Label): 확률 분포 벡터 안에서 가장 높은 확률값을 가진 요소의 위치를 torch.argmax(dim=1)로 찾아내어 정수 형태의 최종 예측 클래스 번호(인덱스)를 확정 짓습니다.

💡 단일 이미지 추론 결과값 상세 분석

* **실행 요약**: 미학습 상태의 모델(`model_0`)에 단일 이미지를 입력하여 반환된 데이터의 흐름과 예측 결과 분석

---

입력 이미지 차원 확인 (`Single image shape`)

* **`torch.Size([1, 3, 64, 64])`**
  * 앞서 `unsqueeze(dim=0)`를 거쳤기 때문에 배치 크기가 `1`로 지정된 4차원 텐서 형태입니다.
  * RGB 컬러 채널(`3`)과 세로·가로 해상도(`64x64`) 격자가 2D CNN 연산에 적합한 표준 규격으로 투입되었음을 나타냅니다.

---

모델 원시 출력값 (`Output logits`)

* **`tensor([[ 0.0208, -0.0020,  0.0095]], device='cuda:0')`**
  * 신경망의 마지막 선형 결합층(`nn.Linear`)을 통과하고 나온 원시 수치 점수(Raw score) 벡터입니다.
  * 정규화되지 않은 값으로, 첫 번째 클래스(인덱스 0)의 점수($0.0208$)가 미세하게 가장 높지만 전체적으로 큰 차이가 없는 박빙의 상태입니다.
  * `device='cuda:0'`을 통해 연산이 GPU 메모리 상에서 성공적으로 처리되었음을 증명합니다.

---

정규화 확률 변환 (`Output prediction probabilities`)

* **`tensor([[0.3371, 0.3295, 0.3333]], device='cuda:0')`**
  * 로짓에 소프트맥스(`torch.softmax`)를 적용하여 모든 클래스의 확률 합이 정확히 1(100%)이 되도록 변환한 수치입니다.
  * **인덱스 0 (피자)**: 33.71%
  * **인덱스 1 (스테이크)**: 32.95%
  * **인덱스 2 (스시)**: 33.33%
  * **해석**: 현재 모델은 아직 학습(Train)을 거치지 않은 무작위 가중치 초기화 상태이므로, 찍기 확률인 약 **33.3%** 부근에서 미세한 난수 값 차이로 요동치고 있음을 보여줍니다.



---

예측 결론 및 실제 정답 비교 (`Output prediction label & Actual label`)

* **`Output prediction label: tensor([0])`**
  * `torch.argmax` 연산이 확률 분포 벡터 내에서 가장 수치가 높은 위치(0번째 인덱스, 33.71%)를 추출하여 최종 예측 결론을 내렸습니다.
* **`Actual label: 0`**
  * 이 데이터의 실제 정답(Ground Truth) 역시 0번 인덱스입니다.
* **최종 결론**: 무작위 가중치 상태의 모델이 우연히 33.71%의 미세한 차이로 정답(0번 클래스)을 맞춘 상황입니다. 모델이 사물의 특징을 온전히 이해하고 맞춘 것이 아니므로, 향후 오차(Loss)를 계산하고 역전파를 수행하는 '모델 학습 과정'이 반드시 필요함을 시사합니다.

### 3.1. 스크립트를 사용하여 모델(TinyVGG) 만들기 (`model_builder.py`)

우리가 작성한 모델 빌드 코드를 다른 곳에서 가져와 사용할 수 있도록(import) 파이썬 스크립트로 변환해 보겠습니다.

In [ ]:
%%writefile going_modular/model_builder.py
"""
CNN Explainer 웹사이트의 TinyVGG 모델을 객체화하기 위한 PyTorch 모델 코드를 포함하는 모듈입니다.
"""
import torch

from torch import nn

class TinyVGG(nn.Module):
  """TinyVGG 신경망 구조를 생성합니다.

  CNN Explainer 웹사이트에 소개된 TinyVGG 구조를 PyTorch로 재현합니다.
  원래 구조는 다음 링크에서 확인할 수 있습니다: https://poloclub.github.io/cnn-explainer/

  Args (매개변수):
    input_shape: 입력 이미지의 채널 수를 나타내는 정수.
    hidden_units: 레이어 사이의 은닉 뉴런(채널) 수를 나타내는 정수.
    output_shape: 최종 출력 뉴런 수를 나타내는 정수.
  """
  def __init__(self, input_shape: int, hidden_units: int, output_shape: int) -> None:
      super().__init__()

      # 1번째 특징 추출 블록: [Conv -> ReLU -> Conv -> ReLU -> MaxPool]
      self.conv_block_1 = nn.Sequential(
          nn.Conv2d(in_channels=input_shape,    # 입력 채널 수 (예: RGB는 3)
                    out_channels=hidden_units,  # 출력 채널 수 (은닉 뉴런 수)
                    kernel_size=3,              # 이미지를 훑는 필터(커널)의 크기 (3x3 격자)
                    stride=1,                   # 필터가 한 번에 이동하는 칸수 (보폭)
                    padding=0),                 # 패딩 없음 ("valid"). 픽셀 외곽에 여백을 주지 않음
          nn.ReLU(),                            # 비선형 활성화 함수
          nn.Conv2d(in_channels=hidden_units,
                    out_channels=hidden_units,
                    kernel_size=3,
                    stride=1,
                    padding=0),
          nn.ReLU(),
          nn.MaxPool2d(kernel_size=2,
                        stride=2) # 가로세로 해상도를 절반으로 압축 (기본 보폭은 커널 크기와 동일한 2)
      )

      # 2번째 특징 추출 블록: 공간 정보를 더 고차원적으로 압축
      self.conv_block_2 = nn.Sequential(
          nn.Conv2d(hidden_units, hidden_units, kernel_size=3, padding=0),
          nn.ReLU(),
          nn.Conv2d(hidden_units, hidden_units, kernel_size=3, padding=0),
          nn.ReLU(),
          nn.MaxPool2d(2) # 크기를 한 번 더 절반으로 줄임
      )

      # 최종 분류기 단계: 압축된 특징들을 일렬로 펼쳐 클래스 점수로 치환
      self.classifier = nn.Sequential(
          nn.Flatten(), # 다차원 텐서를 완전연결층(Linear)에 넣기 위해 1차원 벡터로 평탄화

          # 이 in_features 차원 크기(hidden_units * 13 * 13)는 어떻게 계산되었는가?
          # 네트워크의 각 레이어를 거칠 때마다 입력 이미지 데이터의 형태가 압축되고 변형되기 때문임.
          # (64x64 이미지가 두 번의 Conv와 한 번의 MaxPool 블록 세트를 거치며 최종 13x13 특징 맵으로 축소됨)
          nn.Linear(in_features=hidden_units*13*13, # 평탄화된 전체 특징의 입력 차원 수
                    out_features=output_shape)      # 최종 분류할 정답 클래스의 개수
      )

  def forward(self, x: torch.Tensor):
      # 입력 데이터 x를 정의된 신경망 레이어 순서대로 통과시키는 순방향 연산 수행
      x = self.conv_block_1(x)
      x = self.conv_block_2(x)
      x = self.classifier(x)
      return x
      # return self.classifier(self.block_2(self.block_1(x))) # <- 연산자 퓨전(Operator Fusion)의 이점을 활용하여 한 줄로 압축 표현도 가능함

Writing going_modular/model_builder.py


💡 추가 상세 설명

* **`%%writefile`을 통한 파일 분리의 목적**
  * 주피터 노트북 안에서 모델 구조를 매번 선언하여 사용하는 셀 모드(Cell mode) 방식에서 벗어나기 위함입니다.
  * 노트북 셀 상단에 이 매직 명령어를 적어 실행하면, 훈련 폴더 내에 `model_builder.py`라는 독립된 파이썬 파일이 직접 생성됩니다.
  * 향후 다른 스크립트 파일(`train.py` 등)이나 노트북 환경에서 `from going_modular import model_builder` 형태로 한 줄만 적으면 언제든 똑같은 모델 구조를 깔끔하게 재사용할 수 있습니다.



* **`hidden_units * 13 * 13` 고정 차원의 한계 및 주의점**
  * `in_features` 계산 공식은 입력 이미지 해상도가 전처리 과정에서 **정확히 $64 \times 64$** 크기로 리사이즈되어 들어온다는 가정하에 설계된 수치입니다.
  * 만약 나중에 데이터 전처리 파이프라인에서 이미지 크기를 $128 \times 128$이나 $224 \times 224$ 등 다른 해상도로 변경한다면, 컨볼루션 연산 후 남는 최종 특징 맵의 크기가 $13 \times 13$이 아니게 되므로 완전연결층(`nn.Linear`) 진입 시 `Shape Mismatch` 차원 불일치 에러가 발생하게 됩니다.

In [ ]:
from going_modular import model_builder

In [ ]:
# 분리된 모듈(model_builder)을 활용한 모델 인스턴스 생성 스크립트

import torch

# 1. 앞서 %%writefile로 생성한 going_modular 폴더 내의 model_builder.py 모듈을 임포트
from going_modular import model_builder

# 2. 연산을 가속화할 하드웨어 장치(Device) 설정
device = "cuda" if torch.cuda.is_available() else "cpu"

# 3. model_builder.py 스크립트로부터 TinyVGG 모델 인스턴스 생성 (객체화)
torch.manual_seed(42) # 가중치 초기화의 일관성을 유지하기 위해 무작위 난수 시드 고정

model_1 = model_builder.TinyVGG(
    input_shape=3,                  # 입력 이미지의 컬러 채널 수 (RGB = 3)
    hidden_units=10,                 # 컨볼루션 레이어의 은닉 뉴런(필터) 수
    output_shape=len(class_names)   # 최종 출력 뉴런 수 (클래스 이름 리스트의 길이 = 3)
).to(device)                         # 모델의 가중치들을 지정된 연산 장치(GPU/CPU) 메모리로 전송

# 4. 생성된 model_1의 구조 확인 및 출력
model_1

TinyVGG(
  (conv_block_1): Sequential(
    (0): Conv2d(3, 10, kernel_size=(3, 3), stride=(1, 1))
    (1): ReLU()
    (2): Conv2d(10, 10, kernel_size=(3, 3), stride=(1, 1))
    (3): ReLU()
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (conv_block_2): Sequential(
    (0): Conv2d(10, 10, kernel_size=(3, 3), stride=(1, 1))
    (1): ReLU()
    (2): Conv2d(10, 10, kernel_size=(3, 3), stride=(1, 1))
    (3): ReLU()
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (classifier): Sequential(
    (0): Flatten(start_dim=1, end_dim=-1)
    (1): Linear(in_features=1690, out_features=3, bias=True)
  )
)

💡 TinyVGG 구조 출력 결과(Architecture Output) 상세 분석

* **전체 네트워크 구성**: 2개의 합성곱 블록(`conv_block_1`, `conv_block_2`)과 1개의 최종 분류기 레이어(`classifier`)가 순차적으로 연결된 구조

---

컨볼루션 블록 1 (`conv_block_1`)

* **`(0): Conv2d(3, 10, kernel_size=(3, 3), stride=(1, 1))`**
  * 입력 채널: 3 (RGB 컬러 이미지)
  * 출력 채널: 10 (10개의 가중치 필터로 10차원의 특징 맵 추출)
  * 커널 크기: 3x3 격자 크기로 이미지를 훑음
  * 보폭(Stride): 1칸씩 이동하며 조밀하게 연산
* **`(1) & (3): ReLU()`**
  * 비선형 활성화 함수를 도입하여 모델의 표현력을 높임
* **`(2): Conv2d(10, 10, ...)`**
  * 앞서 추출된 10차원 특징을 그대로 이어받아 사물의 기하학적 패턴을 한 번 더 누적
* **`(4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)`**
  * 2x2 영역 내에서 가장 큰 값만 남기는 맥스 풀링 연산
  * 이미지의 가로세로 크기가 정확히 절반으로 압축됨
  * `ceil_mode=False`: 크기 계산 시 소수점이 발생하면 내림(Floor) 처리하여 자투리 픽셀을 버림

---

컨볼루션 블록 2 (`conv_block_2`)

* **공간 정보의 지속적 압축**: 1번 블록을 거치며 해상도가 대폭 줄어든 텐서를 입력받음
* **채널 크기 유지**: 입력과 출력 채널이 모두 `10`으로 고정됨 (`Conv2d(10, 10)`)
  * 연산 파라미터의 급증을 막으면서, 모델의 수용 영역(Receptive Field)을 넓혀 고차원적 특징 파악

---

최종 분류기 레이어 (`classifier`)

* **`(0): Flatten(start_dim=1, end_dim=-1)`**
  * 배치 차원(0번째 차원)은 그대로 유지함
  * 데이터 입력 표준에 맞추기 위해 이미지 채널과 공간 정보(H, W)가 섞인 다차원 텐서를 1차원 벡터로 일렬 평탄화
* **`(1): Linear(in_features=1690, out_features=3, bias=True)`**
  * `in_features=1690`: 컨볼루션 연산을 모두 거치고 남은 최종 텐서의 크기 ($10\text{채널} \times 13\text{높이} \times 13\text{너비} = 1690$)
  * `out_features=3`: 최종 도출할 예측 클래스 개수 (피자, 스테이크, 스시 총 3종류)
  * `bias=True`: 선형 회귀 연산 시 y절편에 해당하는 편향(Bias) 가중치를 학습 대상에 포함

In [ ]:
# 1. DataLoader에서 이미지와 레이블 한 배치(Batch) 가져오기
# 반복자(Iterator)를 생성한 뒤 next()를 호출하여 첫 번째 미니배치 데이터([32, 3, 64, 64])를 추출함
img_batch, label_batch = next(iter(train_dataloader))

# 2. 배치에서 단일 이미지 추출 및 모델 입력 규격에 맞게 차원 확장(Unsqueeze)
# 배치 내 첫 번째 이미지([3, 64, 64])를 선택한 후, 0번째 차원에 1을 추가하여 [1, 3, 64, 64] 형태로 만듦 (레이블은 정수형 스칼라로 유지)
img_single, label_single = img_batch[0].unsqueeze(dim=0), label_batch[0]
print(f"Single image shape: {img_single.shape}\n")

# 3. 모듈 파일에서 불러온 새 모델(model_1)로 단일 이미지에 대한 순방향 연산(Forward Pass) 진행
model_1.eval() # model_1을 평가(추론) 모드로 전환 (드롭아웃, 배치 정규화 등의 학습용 기능 가동 중지)
with torch.inference_mode(): # 기울기(Gradient) 계산을 비활성화하여 메모리를 절약하고 연산 속도를 최적화함
    pred = model_1(img_single.to(device)) # 이미지를 연산 장치(GPU/CPU)로 전송한 후 모델에 입력하여 가중치 연산 수행

# 4. 연산 과정 출력 및 모델의 출력값(Logits)을 확률(Probs)과 최종 예측 레이블(Label)로 변환
print(f"Output logits:\n{pred}\n") # 모델이 가공 없이 내놓은 원시 예측값(Logits) 출력

# Softmax 함수를 적용하여 3개 클래스의 합이 1(100%)이 되도록 확률 분포값으로 변환
print(f"Output prediction probabilities:\n{torch.softmax(pred, dim=1)}\n")

# argmax 함수를 사용해 확률 분포 중 가장 높은 확률을 가진 인덱스(가장 신뢰도가 높은 클래스 번호)를 추출
print(f"Output prediction label:\n{torch.argmax(torch.softmax(pred, dim=1), dim=1)}\n")
print(f"Actual label:\n{label_single}") # 데이터셋에 기록되어 있던 실제 정답 레이블(Ground Truth) 출력

Single image shape: torch.Size([1, 3, 64, 64])

Output logits:
tensor([[ 0.0208, -0.0020,  0.0095]], device='cuda:0')

Output prediction probabilities:
tensor([[0.3371, 0.3295, 0.3333]], device='cuda:0')

Output prediction label:
tensor([0], device='cuda:0')

Actual label:
0


💡 단일 이미지 추론 결과값 상세 분석

입력 이미지 차원 확인 (`Single image shape`)

* **`torch.Size([1, 3, 64, 64])`**
  * `unsqueeze(dim=0)` 연산으로 인해 맨 앞에 배치 크기 `1`이 추가된 4차원 텐서 형태
  * **차원별 의미**: `[배치 크기(1), 채널 수(3, RGB), 세로 해상도(64), 가로 해상도(64)]`
  * PyTorch 2D CNN 레이어가 요구하는 입력 데이터 표준 규격을 완벽히 충족함

---

모델 원시 출력값 (`Output logits`)

* **`tensor([[ 0.0208, -0.0020,  0.0095]], device='cuda:0')`**
  * 신경망의 최종 출력층(`nn.Linear`)이 뱉어낸 정규화되지 않은 가공 전의 원시 수치 점수(Raw score)
  * 특정 범위 제한($-\infty \sim +\infty$)이 없으며, 인덱스 0번 클래스의 점수($0.0208$)가 미세하게 가장 높음
  * `device='cuda:0'`: 연산이 CPU가 아닌 0번 GPU 메모리 상에서 안전하게 처리되었음을 증명

---

정규화 확률 변환 (`Output prediction probabilities`)

* **`tensor([[0.3371, 0.3295, 0.3333]], device='cuda:0')`**
  * 로짓(Logits) 벡터에 소프트맥스(`torch.softmax`) 함수를 적용하여 전처리한 값
  * 모든 클래스 점수의 총합이 정확히 `1.0 (100%)`이 되도록 변환하여 직관적인 확률 분포로 표현
  * **클래스별 신뢰도 결과**:
    * 인덱스 0 (Pizza): **33.71%**
    * 인덱스 1 (Steak): **32.95%**
    * 인덱스 2 (Sushi): **33.33%**
  * **상태 분석**: 현재 모델은 학습을 전혀 거치지 않은 가중치 초기화 상태이므로, 단순 찍기 확률인 3분의 1(약 33.3%) 부근에서 난수 값 차이로 비등비등하게 요동치고 있음

---

예측 결론 및 실제 정답 비교 (`Output prediction label & Actual label`)

* **`Output prediction label: tensor([0])`**
  * `torch.argmax` 함수가 확률 분포 벡터 내에서 가장 높은 수치를 기록한 위치(0번 인덱스, 33.71%)를 최종 판정값으로 채택
* **`Actual label: 0`**
  * 데이터셋에 명시되어 있던 실제 정답(Ground Truth) 레이블 또한 0번 인덱스임
* **최종 결론**: 무작위 초기화 상태의 모델이 우연히 33.71%의 미세한 점수 차이로 정답(0번 클래스)을 맞춘 상황. 사물의 특징을 이해한 결과가 아니므로, 향후 오차(Loss)를 계산하고 역전파(Backpropagation)를 가동하는 본 가공 학습 과정이 필수적임을 시사함

## 4. `train_step()` 및 `test_step()` 함수와 이들을 결합할 `train()` 함수 만들기

> 177. 모델 훈련 코드를 파이썬 스크립트로 변환하기: Turning Our Model Training Code into a Python Script

* **기존 코드의 재사용**
  * 코드를 처음부터 다시 작성하는 대신, [노트북 04](https://www.learnpytorch.io/04_pytorch_custom_datasets/#75-create-train-test-loop-functions)에서 완성했던 `train_step()`과 `test_step()` 함수를 그대로 재사용함
  * 여러 개의 에포크(Epoch) 동안 두 함수를 반복 실행하며 통합 제어하던 `train()` 함수도 동일하게 재사용 가능

* **이전 코드와의 유일한 차이점**
  * 함수 내부에 기능 설명, 매개변수(Args), 반환값(Returns) 정보를 명시하는 독스트링(Docstring)이 추가됨
  * 해당 독스트링은 규격화된 [Google의 파이썬 함수 및 메서드 스타일 가이드](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods)를 엄격히 준수하여 작성됨

* **다음 단계**
  * 가장 먼저 모듈의 첫 번째 부품이 될 `train_step()` 함수 정의부터 시작함

In [ ]:

from typing import Tuple

def train_step(model: torch.nn.Module,
               dataloader: torch.utils.data.DataLoader,
               loss_fn: torch.nn.Module,
               optimizer: torch.optim.Optimizer,
               device: torch.device) -> Tuple[float, float]:
  """단일 에포크(single epoch) 동안 PyTorch 모델을 훈련합니다.

  대상 PyTorch 모델을 훈련 모드(training mode)로 전환한 다음,
  필요한 모든 훈련 단계(순방향 연산, 손실 계산, 옵티마이저 가중치 업데이트)를 실행합니다.

  Args:
    model: 훈련할 PyTorch 모델.
    dataloader: 모델을 훈련하는 데 사용할 데이터가 담긴 DataLoader 인스턴스.
    loss_fn: 최소화할 PyTorch 손실 함수 (Loss function).
    optimizer: 손실 함수를 줄이는 데 도움을 줄 PyTorch 옵티마이저 (Optimizer).
    device: 연산을 수행할 대상 장치 (예: "cuda" 또는 "cpu").

  Returns:
    훈련 손실(train_loss)과 훈련 정확도(train_accuracy) 메트릭을 담은 튜플.
    형태: (train_loss, train_accuracy). 예시: (0.1112, 0.8743)
  """
  # 1. 모델을 훈련 모드로 설정
  # 드롭아웃(Dropout)이나 배치 정규화(Batch Normalization) 등 학습 시에만 필요한 기능들을 활성화함
  model.train()

  # 2. 에포크별 총 손실 및 정확도를 누적할 변수 초기화
  train_loss, train_acc = 0, 0

  # 3. 데이터 로더의 미니배치(Batch) 단위로 루프 돌기
  for batch, (X, y) in enumerate(dataloader):
      # 입력 데이터(X: 이미지)와 정답 레이블(y)을 지정된 하드웨어 장치(GPU/CPU)로 전송
      X, y = X.to(device), y.to(device)

      # 3-1. 순방향 연산 (Forward pass)
      # 모델에 데이터를 입력하여 예측 스코어(Logits) 벡터를 도출함
      y_pred = model(X)

      # 3-2. 손실 계산 및 누적 (Calculate & accumulate loss)
      # 예측값과 정답 간의 오차를 계산하고, item()을 통해 파이썬 스칼라 값으로 변환하여 누적 변수에 더함
      loss = loss_fn(y_pred, y)
      train_loss += loss.item()

      # 3-3. 옵티마이저 변화도 초기화 (Optimizer zero grad)
      # 이전 미니배치 연산에서 계산되어 누적되어 있던 기울기(Gradient) 값을 0으로 리셋함
      optimizer.zero_grad()

      # 3-4. 역전파 수행 (Loss backward)
      # 가중치 손실에 대한 기울기를 역방향으로 전파하며 자동으로 계산하고 각 파라미터에 저장함
      loss.backward()

      # 3-5. 매개변수 업데이트 (Optimizer step)
      # 계산된 기울기를 바탕으로 모델의 가중치(Weights)와 편향(Biases)을 최적화 알고리즘에 맞게 갱신함
      optimizer.step()

      # 3-6. 배치별 정확도 계산 및 누적
      # 소프트맥스로 확률 분포를 만들고 argmax로 최고 신뢰도 인덱스를 뽑아낸 뒤, 정답과 비교하여 배치 내 평균 정확도를 누적
      y_pred_class = torch.argmax(torch.softmax(y_pred, dim=1), dim=1)
      train_acc += (y_pred_class == y).sum().item()/len(y_pred)

  # 4. 전체 배치의 평균 메트릭 계산
  # 누적된 값을 총 배치 개수(len(dataloader))로 나누어 해당 에포크의 평균 손실과 정확도를 도출함
  train_loss = train_loss / len(dataloader)
  train_acc = train_acc / len(dataloader)

  return train_loss, train_acc

Now we'll do `test_step()`.

In [ ]:
def test_step(model: torch.nn.Module,
              dataloader: torch.utils.data.DataLoader,
              loss_fn: torch.nn.Module,
              device: torch.device) -> Tuple[float, float]:
  """단일 에포크(single epoch) 동안 PyTorch 모델을 검증(테스트)합니다.

  대상 PyTorch 모델을 평가 모드("eval" mode)로 전환한 다음,
  테스트 데이터셋에 대해 순방향 연산(Forward pass)을 수행합니다.

  Args:
    model: 검증할 PyTorch 모델.
    dataloader: 모델을 검증하는 데 사용할 데이터가 담긴 DataLoader 인스턴스.
    loss_fn: 테스트 데이터에 대한 오차를 계산할 PyTorch 손실 함수.
    device: 연산을 수행할 대상 장치 (예: "cuda" 또는 "cpu").

  Returns:
    테스트 손실(test_loss)과 테스트 정확도(test_accuracy) 메트릭을 담은 튜플.
    형태: (test_loss, test_accuracy). 예시: (0.0223, 0.8985)
  """
  # 1. 모델을 평가(추론) 모드로 설정
  # 드롭아웃(Dropout)이나 배치 정규화(Batch Normalization) 등 학습 시에만 필요한 기능을 비활성화함
  model.eval()

  # 2. 에포크별 총 테스트 손실 및 정확도를 누적할 변수 초기화
  test_loss, test_acc = 0, 0

  # 3. 추론 컨텍스트 매니저 활성화
  # 기울기(Gradient) 계산을 차단하여 메모리 오버헤드를 막고 연산 속도를 가속화함
  with torch.inference_mode():
      # 테스트 데이터 로더의 미니배치(Batch) 단위로 루프 돌기
      for batch, (X, y) in enumerate(dataloader):
          # 입력 데이터(X: 이미지)와 정답 레이블(y)을 지정된 하드웨어 장치(GPU/CPU)로 전송
          X, y = X.to(device), y.to(device)

          # 3-1. 순방향 연산 (Forward pass)
          # 테스트 데이터를 모델에 입력하여 예측 스코어(Logits) 벡터를 도출함
          test_pred_logits = model(X)

          # 3-2. 테스트 손실 계산 및 누적 (Calculate and accumulate loss)
          # 예측값과 정답 간의 오차를 계산하고, item()을 통해 파이썬 스칼라 값으로 변환하여 누적 변문에 더함
          loss = loss_fn(test_pred_logits, y)
          test_loss += loss.item()

          # 3-3. 테스트 정확도 계산 및 누적
          # argmax(dim=1)를 사용해 소프트맥스 통과 전 로짓 단계에서 바로 최고 점수의 인덱스를 추출함
          test_pred_labels = test_pred_logits.argmax(dim=1)
          # 예측한 인덱스와 실제 정답(y)을 비교하여 배치 내 평균 정확도를 구한 후 누적 변수에 합산
          test_acc += ((test_pred_labels == y).sum().item()/len(test_pred_labels))

  # 4. 전체 배치의 평균 검증 메트릭 계산
  # 누적된 값을 총 배치 개수(len(dataloader))로 나누어 해당 검증 에포크의 평균 손실과 정확도를 도출함
  test_loss = test_loss / len(dataloader)
  test_acc = test_acc / len(dataloader)

  return test_loss, test_acc

훈련 단계(train_step)와의 결정적 차이점

- 오차 역전파 및 가중치 갱신 생략: optimizer.zero_grad(), loss.backward(), optimizer.step() 구조가 전혀 존재하지 않음 (순수 성능 평가 목적)

- 컨텍스트 제한: with torch.inference_mode() 내부에서 연산이 수행되므로 불필요한 연산 그래프 텐서가 누적되지 않음

argmax(dim=1)의 최적화 효율성

- 예측 레이블을 뽑기 위해 굳이 torch.softmax()를 거쳐 연산 자원을 소모하지 않고, 원시 로짓(test_pred_logits) 상태에서 곧바로 최댓값 인덱스를 추출함

- 로짓 값의 대소 관계 순위와 소프트맥스 확률값의 대소 관계 순위는 완벽히 일치하기 때문에 검증 속도를 높이는 최적화 패턴으로 자주 사용됨

And we'll combine `train_step()` and `test_step()` into `train()`.

In [ ]:
from typing import Dict, List
from tqdm.auto import tqdm

def train(model: torch.nn.Module,
          train_dataloader: torch.utils.data.DataLoader,
          test_dataloader: torch.utils.data.DataLoader,
          optimizer: torch.optim.Optimizer,
          loss_fn: torch.nn.Module,
          epochs: int,
          device: torch.device) -> Dict[str, List[float]]:
  """PyTorch 모델을 학습 및 검증(테스트)합니다.

  지정된 에포크(Epoch) 수만큼 대상 PyTorch 모델을 train_step()과 test_step()
  함수에 통과시키며, 동일한 에포크 루프 안에서 모델의 학습과 테스트를 함께 진행합니다.

  전체 과정 동안의 평가 메트릭을 계산, 출력 및 저장합니다.

  Args:
    model: 훈련 및 테스트를 진행할 PyTorch 모델.
    train_dataloader: 모델을 학습시킬 데이터가 담긴 DataLoader 인스턴스.
    test_dataloader: 모델을 검증할 데이터가 담긴 DataLoader 인스턴스.
    optimizer: 손실 함수를 최소화하는 데 사용할 PyTorch 옵티마이저.
    loss_fn: 두 데이터셋에서 오차(손실)를 계산할 PyTorch 손실 함수.
    epochs: 모델을 몇 에포크 동안 학습시킬지 지정하는 정수.
    device: 연산을 수행할 대상 장치 (예: "cuda" 또는 "cpu").

  Returns:
    훈련 및 테스트 손실, 그리고 훈련 및 테스트 정확도 메트릭이 담긴 딕셔너리.
    각 메트릭 리스트는 에포크별 결과값을 순서대로 저장함.
    형태: {train_loss: [...], train_acc: [...], test_loss: [...], test_acc: [...]}
    예시 (epochs=2 인 경우):
         {train_loss: [2.0616, 1.0537],
          train_acc: [0.3945, 0.3945],
          test_loss: [1.2641, 1.5706],
          test_acc: [0.3400, 0.2973]}
  """
  # 1. 메트릭 결과를 누적 저장할 빈 결과 딕셔너리 생성
  results = {"train_loss": [],
      "train_acc": [],
      "test_loss": [],
      "test_acc": []
  }

  # 2. 지정된 에포크 횟수만큼 학습 및 검증 루프 반복 실행
  # tqdm을 사용하여 콘솔에 훈련 진행 상황을 나타내는 프로그레스 바(Progress Bar)를 출력함
  for epoch in tqdm(range(epochs)):
      # 2-1. 단일 에포크 훈련 진행 (가중치 업데이트 발생)
      train_loss, train_acc = train_step(model=model,
                                          dataloader=train_dataloader,
                                          loss_fn=loss_fn,
                                          optimizer=optimizer,
                                          device=device)
      # 2-2. 단일 에포크 검증 진행 (가중치 고정 상태로 성능만 측정)
      test_loss, test_acc = test_step(model=model,
          dataloader=test_dataloader,
          loss_fn=loss_fn,
          device=device)

      # 2-3. 현재 에포크의 성적(손실 및 정확도)을 소수점 4자리까지 포맷팅하여 콘솔에 출력
      print(
          f"Epoch: {epoch+1} | "
          f"train_loss: {train_loss:.4f} | "
          f"train_acc: {train_acc:.4f} | "
          f"test_loss: {test_loss:.4f} | "
          f"test_acc: {test_acc:.4f}"
      )

      # 2-4. 계산된 에포크별 메트릭 값을 결과 딕셔너리 리스트에 각각 추가(Append)
      results["train_loss"].append(train_loss)
      results["train_acc"].append(train_acc)
      results["test_loss"].append(test_loss)
      results["test_acc"].append(test_acc)

  # 3. 모든 에포크가 종료되면 최종 학습 기록 데이터가 담긴 딕셔너리를 반환
  return results

- 통합 마스터 루프의 역할: 앞서 개별 부품으로 정의했던 train_step()과 test_step()을 하나의 에포크 사이클로 묶어주는 엔진 역할을 수행합니다.

- 시각적 모니터링 (tqdm): 파이썬 빌트인 range() 대신 tqdm(range())을 사용하여 대규모 학습 시 에포크당 걸리는 속도와 예상 완료 시간을 실시간 바 형태로 보여줍니다.

- 결과 히스토리 자산화: 에포크별 메트릭을 results 딕셔너리에 지속적으로 누적 반환하므로, 학습 종료 후 Matplotlib 등의 라이브러리를 활용해 손실 곡선(Loss Curve) 및 정확도 그래프를 시각화할 때 핵심 기초 자료로 활용됩니다.

### 4.1 Training function을 스크립트로 변환하기 (`engine.py`)

In [ ]:
%%writefile going_modular/engine.py
"""
PyTorch 모델을 훈련(Training)하고 검증(Testing)하는 데 필요한 핵심 핵심 함수들을 포함하는 모듈입니다.
"""
from typing import Dict, List, Tuple

import torch

from tqdm.auto import tqdm

def train_step(model: torch.nn.Module,
               dataloader: torch.utils.data.DataLoader,
               loss_fn: torch.nn.Module,
               optimizer: torch.optim.Optimizer,
               device: torch.device) -> Tuple[float, float]:
  """단일 에포크(Single epoch) 동안 PyTorch 모델을 학습시킵니다.

  대상 모델을 학습 모드로 설정하고 순방향 연산, 손실 계산, 역전파,
  그리고 가중치 최적화 단계를 순차적으로 실행합니다.

  Args:
    model: 학습시킬 PyTorch 모델.
    dataloader: 학습 데이터셋이 담긴 DataLoader 인스턴스.
    loss_fn: 오차를 최소화하기 위한 PyTorch 손실 함수.
    optimizer: 가중치 갱신을 수행할 PyTorch 옵티마이저.
    device: 연산이 수행될 장치 (예: "cuda" 또는 "cpu").

  Returns:
    (평균 훈련 손실, 평균 훈련 정확도) 형태의 튜플. 예: (0.1112, 0.8743)
  """
  # 모델을 훈련 모드로 전환 (드롭아웃, 배치 정규화 등 활성화)
  model.train()

  # 에포크 단위 손실 및 정확도 누적 변수 초기화
  train_loss, train_acc = 0, 0

  # 미니배치 단위 루프 실행
  for batch, (X, y) in enumerate(dataloader):
      # 입력 데이터(X)와 정답(y)을 해당 디바이스(GPU/CPU) 메모리로 전송
      X, y = X.to(device), y.to(device)

      # 1. 순방향 연산 (Forward pass): 모델 예측값 계산
      y_pred = model(X)

      # 2. 배치 손실 계산 및 누적 (파이썬 기본 스칼라 값으로 변환하여 더함)
      loss = loss_fn(y_pred, y)
      train_loss += loss.item()

      # 3. 옵티마이저 내 기존 가중치 변화도(Gradient)를 0으로 초기화
      optimizer.zero_grad()

      # 4. 오차 역전파 (Backpropagation): 변화도 계산
      loss.backward()

      # 5. 가중치 매개변수 업데이트 (Optimizer step)
      optimizer.step()

      # 배치별 정확도 계산 및 누적
      y_pred_class = torch.argmax(torch.softmax(y_pred, dim=1), dim=1)
      train_acc += (y_pred_class == y).sum().item()/len(y_pred)

  # 전체 배치 개수로 나누어 에포크 평균 손실 및 정확도 산출
  train_loss = train_loss / len(dataloader)
  train_acc = train_acc / len(dataloader)
  return train_loss, train_acc

def test_step(model: torch.nn.Module,
              dataloader: torch.utils.data.DataLoader,
              loss_fn: torch.nn.Module,
              device: torch.device) -> Tuple[float, float]:
  """단일 에포크(Single epoch) 동안 PyTorch 모델의 성능을 검증합니다.

  대상 모델을 평가 모드로 전환하고 테스트 데이터셋에 대해 순방향 연산을 수행하여 성능을 측정합니다.

  Args:
    model: 검증할 PyTorch 모델.
    dataloader: 검증 데이터셋이 담긴 DataLoader 인스턴스.
    loss_fn: 검증 오차를 계산할 PyTorch 손실 함수.
    device: 연산이 수행될 장치 (예: "cuda" 또는 "cpu").

  Returns:
    (평균 테스트 손실, 평균 테스트 정확도) 형태의 튜플. 예: (0.0223, 0.8985)
  """
  # 모델을 평가 모드로 전환 (드롭아웃, 배치 정규화 등 고정)
  model.eval()

  # 에포크 단위 검증 손실 및 정확도 누적 변수 초기화
  test_loss, test_acc = 0, 0

  # 기울기 추적을 차단하여 메모리 절약 및 추론 속도 최적화
  with torch.inference_mode():
      # 검증용 미니배치 단위 루프 실행
      for batch, (X, y) in enumerate(dataloader):
          # 데이터를 하드웨어 장치(GPU/CPU) 메모리로 전송
          X, y = X.to(device), y.to(device)

          # 1. 순방향 연산 (Forward pass): 원시 예측값(Logits) 도출
          test_pred_logits = model(X)

          # 2. 검증 손실 계산 및 누적
          loss = loss_fn(test_pred_logits, y)
          test_loss += loss.item()

          # 검증 정확도 계산 및 누적 (로짓 상태에서 바로 argmax 적용)
          test_pred_labels = test_pred_logits.argmax(dim=1)
          test_acc += ((test_pred_labels == y).sum().item()/len(test_pred_labels))

  # 전체 배치 개수로 나누어 검증 에포크 평균 손실 및 정확도 산출
  test_loss = test_loss / len(dataloader)
  test_acc = test_acc / len(dataloader)
  return test_loss, test_acc


def train(model: torch.nn.Module,
          train_dataloader: torch.utils.data.DataLoader,
          test_dataloader: torch.utils.data.DataLoader,
          optimizer: torch.optim.Optimizer,
          loss_fn: torch.nn.Module,
          epochs: int,
          device: torch.device) -> Dict[str, List[float]]:
  """전체 에포크 루프를 돌며 PyTorch 모델을 학습 및 테스트(검증)합니다.

  동일한 에포크 반복문 안에서 train_step()과 test_step()을 순서대로 실행하며,
  매 에포크마다의 성능 메트릭을 콘솔에 출력하고 딕셔너리에 기록합니다.

  Args:
    model: 학습 및 검증을 진행할 PyTorch 모델.
    train_dataloader: 학습용 DataLoader.
    test_dataloader: 검증용 DataLoader.
    optimizer: 학습 오차를 줄여나갈 PyTorch 옵티마이저.
    loss_fn: 두 단계 모두에서 사용할 PyTorch 손실 함수.
    epochs: 총 몇 번의 주기를 반복하여 학습시킬지 결정하는 정수.
    device: 연산 장치 ("cuda" 또는 "cpu").

  Returns:
    모든 에포크의 히스토리 메트릭이 리스트 형태로 누적된 결과 딕셔너리.
    형태: {"train_loss": [...], "train_acc": [...], "test_loss": [...], "test_acc": [...]}
  """
  # 에포크별 전체 학습 추이를 기록할 저장소 딕셔너리 정의
  results = {"train_loss": [],
      "train_acc": [],
      "test_loss": [],
      "test_acc": []
  }

  # tqdm 프로그레스 바를 동반하여 지정된 에포크 수만큼 반복문 실행
  for epoch in tqdm(range(epochs)):
      # 1 에포크 분량의 가중치 학습 프로세스 수행
      train_loss, train_acc = train_step(model=model,
                                          dataloader=train_dataloader,
                                          loss_fn=loss_fn,
                                          optimizer=optimizer,
                                          device=device)
      # 1 에포크 분량의 모델 검증 프로세스 수행
      test_loss, test_acc = test_step(model=model,
          dataloader=test_dataloader,
          loss_fn=loss_fn,
          device=device)

      # 현재 진행 주기와 메트릭 결과물들을 포맷팅하여 콘솔 화면에 실시간 출력
      print(
          f"Epoch: {epoch+1} | "
          f"train_loss: {train_loss:.4f} | "
          f"train_acc: {train_acc:.4f} | "
          f"test_loss: {test_loss:.4f} | "
          f"test_acc: {test_acc:.4f}"
      )

      # 반환할 딕셔너리의 각 리스트 자리에 현재 에포크 결과값을 누적(Append)
      results["train_loss"].append(train_loss)
      results["train_acc"].append(train_acc)
      results["test_loss"].append(test_loss)
      results["test_acc"].append(test_acc)

  # 모든 루프가 종료되면 최종 리포트 데이터가 누적된 딕셔너리 반환
  return results

Writing going_modular/engine.py


%%writefile going_modular/engine.py 매직 커맨드를 통해 파이토치 학습 파이프라인의 핵심인 훈련 엔진을 독자적인 외부 파이썬 스크립트 파일로 완전히 분리·저장하는 코드입니다.

이 모듈화 작업을 거치고 나면 다음 실전 프로젝트부터 지저분한 메인 루프 구현부를 작성할 필요 없이 from going_modular.engine import train 구문 한 줄 만으로 정교한 인공지능 학습 알고리즘을 즉시 끌어다 쓸 수 있습니다.

In [ ]:
from going_modular import engine

# engine.train()

## 5. 모델을 저장하는 함수 만들기

> 178. 모델을 저장하는 유틸리티 함수를 파이썬 스크립트로 변환하기: Turning Our Utility Function to Save a Model into a Python Script


학습이 완료된 PyTorch 모델의 가중치(Weights) 파라미터를 특정 디렉토리(폴더)에 파일 형태로 안전하게 보관할 수 있도록 전용 함수를 구성하는 단계

In [ ]:
from pathlib import Path

def save_model(model: torch.nn.Module,
               target_dir: str,
               model_name: str):
  """대상 PyTorch 모델을 지정된 디렉토리에 저장합니다.

  Args:
    model: 저장할 대상 PyTorch 모델.
    target_dir: 모델을 저장할 목적지 폴더 경로.
    model_name: 저장될 모델의 파일 이름. 확장자로 ".pth" 또는 ".pt"를
      반드시 포함해야 합니다.

  Example usage (사용 예시):
    save_model(model=model_0,
               target_dir="models",
               model_name="05_going_modular_tingvgg_model.pth")
  """
  # 1. 대상 디렉토리(폴더) 경로 객체 생성 및 폴더 생성
  target_dir_path = Path(target_dir)
  # parents=True: 상위 폴더가 없으면 상위 폴더까지 함께 생성함
  # exist_ok=True: 이미 해당 폴더가 존재하더라도 에러를 발생시키지 않고 무시함
  target_dir_path.mkdir(parents=True,
                        exist_ok=True)

  # 2. 최종 모델 저장 경로 결정 및 파일 확장자 검증
  # assert 조건을 만족하지 않으면 지정된 에러 메시지를 출력하며 프로그램이 중단됨
  assert model_name.endswith(".pth") or model_name.endswith(".pt"), "model_name은 반드시 '.pt' 또는 '.pth'로 끝나야 합니다."
  # 슬래시(/) 연산자를 사용해 폴더 경로와 파일 이름을 안전하게 결합 (OS별 경로 구분자 자동 처리)
  model_save_path = target_dir_path / model_name

  # 3. 모델의 가중치 상태 사전(state_dict)을 파일로 최종 저장
  print(f"[INFO] Saving model to: {model_save_path}")
  # 전체 모델 객체 대신 가중치 데이터만 딕셔너리 형태로 추출(state_dict())하여 효율적으로 백업
  torch.save(obj=model.state_dict(),
             f=model_save_path)

안전한 디렉토리 제어 (Path.mkdir)

- 코드가 실행되는 환경에 해당 저장용 폴더가 미리 파여 있지 않더라도 예외 에러(FileNotFoundError) 없이 런타임 중에 디렉토리를 자동 빌드합니다.

확장자 유효성 예외 처리 (assert)

- 사용자가 파일명 뒤에 실수로 .txt나 .csv를 붙이거나 확장자를 누락하는 실수를 원천 차단하여, PyTorch가 인식할 수 있는 정상 포맷(.pt / .pth) 저장을 강제합니다.

state_dict() 기반 저장의 이점

- 모델 구조 전체를 직렬화하여 저장하는 방식에 비해 파일 용량이 대폭 압축됩니다.

- 다른 파이썬 파일이나 환경에서 모델의 구조 클래스만 똑같이 선언해 두면, 이 가중치 딕셔너리 파일을 레이어에 쏙 끼워 맞춰서(load_state_dict()) 깨끗하게 복원할 수 있어 이식성이 매우 뛰어납니다.

### 5.1 유틸리티 함수들을 모아둔 `utils.py` 파일 만들기

파이썬 프로그래밍에서 "utils"라는 이름은 보통 프로젝트 전반에서 공통으로 쓰이는 다양한 유틸리티(도구성) 함수들을 한데 모아 관리하는 전용 공간으로 예약되어 사용됨

현재 단계에서 우리가 가진 유틸리티 함수는 모델을 저장하는 함수인 save_model() 단 하나뿐임

하지만 프로젝트의 규모가 커지고 코드가 복잡해질수록, 이미지 시각화, 로그 기록, 체크포인트 로드 등 더 많은 유틸리티 함수가 추가될 가능성이 높음

In [ ]:
%%writefile going_modular/utils.py
"""
File containing various utility functions for PyTorch model training.
"""
import torch

from pathlib import Path

def save_model(model: torch.nn.Module,
               target_dir: str,
               model_name: str):
  """대상 PyTorch 모델의 가중치를 지정된 디렉토리에 안전하게 저장합니다.

  Args:
    model: 저장할 대상 PyTorch 모델 객체.
    target_dir: 모델 파일을 저장할 목적지 폴더 경로 문자열.
    model_name: 저장할 파일의 이름. 확장자로 반드시 '.pt' 또는 '.pth'를
      포함해야 합니다.

  Example usage:
    save_model(model=model_0,
               target_dir="models",
               model_name="05_going_modular_tingvgg_model.pth")
  """
  # 1. 목적지 디렉토리 경로를 Path 객체로 변환하고 폴더가 없으면 자동 생성
  # parents=True: 중간 단계의 상위 폴더가 없다면 함께 생성함
  # exist_ok=True: 이미 해당 폴더가 생성되어 있어도 에러 없이 진행함
  target_dir_path = Path(target_dir)
  target_dir_path.mkdir(parents=True,
                        exist_ok=True)

  # 2. 파일 확장자 유효성 검사 및 최종 절대/상대 경로 결합
  # 지정된 확장자가 아닐 경우 AssertionError를 발생시키며 프로그램을 안전하게 중단시킴
  assert model_name.endswith(".pth") or model_name.endswith(".pt"), "model_name은 반드시 '.pt' 또는 '.pth'로 끝나야 합니다."
  model_save_path = target_dir_path / model_name

  # 3. 모델의 가중치 데이터(state_dict)만을 추출하여 지정된 경로에 최종 파일로 바이너리 저장
  print(f"[INFO] Saving model to: {model_save_path}")
  torch.save(obj=model.state_dict(),
             f=model_save_path)

Writing going_modular/utils.py


## 6. 모델 훈련, 평가 및 저장하기

> 179. 코드 한 줄로 모델을 훈련시키는 훈련 스크립트 만들기: Creating a Training Script to Train Our Model in One Line of Code

앞 단계에서 우리가 열심히 쪼개고 분리해 두었던 여러 기능성 함수들(data_setup, model_builder, engine, utils)을 총동원하여, 모델의 전체 훈련-검증-저장 프로세스를 단 한 번에 수행하는 단계입니다.


In [ ]:
# 1. 재현성을 위한 무작위 시드(Seed) 설정
# CPU와 GPU 연산의 무작위 난수 생성기 시드를 고정하여, 언제 어디서 코드를 다시 돌려도 똑같은 실험 결과가 나오도록 보장함
torch.manual_seed(42)
torch.cuda.manual_seed(42)

# 2. 하이퍼파라미터 정의
NUM_EPOCHS = 5 # 전체 데이터셋을 총 5번 반복하여 학습하도록 에포크 설정

# 3. TinyVGG 모델 인스턴스 초기화 및 장치 할당
# 앞서 정의한 구조 클래스를 호출하고 .to(device)를 통해 지정한 연산 하드웨어(GPU/CPU) 메모리로 모델 가중치를 전송
model_0 = TinyVGG(input_shape=3, # 입력 이미지의 색상 채널 수 (RGB 컬러이므로 3)
                  hidden_units=10, # 은닉층 컨볼루션 레이어들의 커널(필터) 개수
                  output_shape=len(train_data.classes)).to(device) # 최종 분류할 클래스 개수 (3개 종류)

# 4. 손실 함수 및 최적화 알고리즘(옵티마이저) 설정
loss_fn = nn.CrossEntropyLoss() # 다중 클래스 분류를 위한 교차 엔트로피 손실 함수 정의
optimizer = torch.optim.Adam(params=model_0.parameters(), lr=0.001) # Adam 옵티마이저에 모델 파라미터와 학습률(Learning Rate) 등록

# 5. 학습 소요 시간 측정을 위한 타이머 가동
from timeit import default_timer as timer
start_time = timer() # 현재 시점의 타임스탬프 기록

# 6. 모듈 통합 train() 함수를 통한 모델 학습 및 검증 파이프라인 가동
# 지정된 에포크 동안 train_step과 test_step이 내부에서 순차적으로 맞물려 실행되며 결과 로그를 저장함
model_0_results = train(model=model_0,
                        train_dataloader=train_dataloader,
                        test_dataloader=test_dataloader,
                        optimizer=optimizer,
                        loss_fn=loss_fn,
                        epochs=NUM_EPOCHS,
                        device=device)

# 7. 타이머 종료 및 총 훈련 소요 시간 출력
end_time = timer() # 종료 시점의 타임스탬프 기록
print(f"[INFO] Total training time: {end_time-start_time:.3f} seconds") # 소요된 순수 초(seconds) 단위를 소수점 3자리까지 출력

# 8. 유틸리티 함수를 이용한 모델 최종 본 가중치 백업
# 학습이 완료된 최적 가중치 상태(state_dict)를 "models" 디렉토리 안에 지정된 이름의 .pth 파일로 안전하게 직렬화 저장함
save_model(model=model_0,
           target_dir="models",
           model_name="05_going_modular_cell_mode_tinyvgg_model.pth")

  0%|          | 0/5 [00:00<?, ?it/s]

Epoch: 1 | train_loss: 1.1063 | train_acc: 0.3047 | test_loss: 1.0983 | test_acc: 0.3011
Epoch: 2 | train_loss: 1.0998 | train_acc: 0.3281 | test_loss: 1.0697 | test_acc: 0.5417
Epoch: 3 | train_loss: 1.0869 | train_acc: 0.4883 | test_loss: 1.0807 | test_acc: 0.4924
Epoch: 4 | train_loss: 1.0844 | train_acc: 0.4023 | test_loss: 1.0606 | test_acc: 0.5833
Epoch: 5 | train_loss: 1.0663 | train_acc: 0.4141 | test_loss: 1.0657 | test_acc: 0.5644
[INFO] Total training time: 6.155 seconds
[INFO] Saving model to: models/05_going_modular_cell_mode_tinyvgg_model.pth


### 6.1 델 학습, 평가 및 저장 (스크립트 모드) -> `train.py`

우리가 작성한 다른 모든 코드 스크립트들을 활용하여 PyTorch 모델을 학습시킬 수 있도록 train.py라는 파일을 만들어 보겠습니다.

기본적으로 우리가 원하는 것은 04번 노트북(notebook)에서 했던 기능을 단 한 줄로 똑같이 구현하는 것입니다...

In [ ]:
%%writefile going_modular/train.py
"""
Trains a PyTorch image classification model using device-agnostic code.
"""
import os
import torch

from torchvision import transforms
from timeit import default_timer as timer

# 앞서 작성하여 저장해 둔 커스텀 모듈 파일(.py)들을 임포트
import data_setup, engine, model_builder, utils

# 1. 하이퍼파라미터(Hyperparameters) 설정
NUM_EPOCHS = 5          # 전체 데이터셋을 반복 학습할 횟수
BATCH_SIZE = 32         # 한 번에 모델에 입력할 데이터 개수
HIDDEN_UNITS = 10       # CNN 은닉층의 채널(필터) 개수
LEARNING_RATE = 0.001   # 가중치를 업데이트할 때 반영할 학습률

# 2. 데이터 디렉토리 경로 정의
train_dir = "data/pizza_steak_sushi/train"  # 학습 데이터셋 경로
test_dir = "data/pizza_steak_sushi/test"    # 검증(테스트) 데이터셋 경로

# 3. 연산 장치 할당 (CUDA GPU 사용이 가능하면 가속기를, 불가능하면 CPU를 자동 지정)
device = "cuda" if torch.cuda.is_available() else "cpu"

# 4. 이미지 전처리(Transform) 파이프라인 생성
# 모든 이미지를 64x64 크기로 리사이즈하고, PyTorch 연산이 가능하도록 텐서([0.0, 1.0] 범위)로 변환
data_transform = transforms.Compose([
                                     transforms.Resize((64, 64)),
                                     transforms.ToTensor()
])

# 5. data_setup 모듈을 활용해 데이터로더(DataLoader) 구축 및 클래스 이름 추출
# 앞서 설정한 경로, 전처리 툴, 배치 크기를 전달하여 훈련/테스트용 데이터로더와 클래스 목록(['pizza', 'steak', 'sushi'])을 반환받음
train_dataloader, test_dataloader, class_names = data_setup.create_dataloaders(train_dir=train_dir,
                                                                               test_dir=test_dir,
                                                                               transform=data_transform,
                                                                               batch_size=BATCH_SIZE)

# 6. model_builder 모듈을 활용해 TinyVGG 모델 인스턴스 생성 및 연산 장치 이동
# 입력 채널(RGB=3), 은닉층 유닛 수, 출력 클래스 개수(3)를 지정하여 모델을 빌드하고 지정된 하드웨어(GPU/CPU)에 올림
model = model_builder.TinyVGG(input_shape=3,
                              hidden_units=HIDDEN_UNITS,
                              output_shape=len(class_names)).to(device)

# 7. 다중 분류를 위한 손실 함수(Loss function) 및 가중치 업데이트용 Adam 옵티마이저(Optimizer) 설정
loss_fn = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(),
                             lr=LEARNING_RATE)

# 8. 총 훈련 소요 시간 측정을 위한 타이머 시작
start_time = timer()

# 9. engine 모듈의 train() 함수를 호출하여 통합 모델 학습 및 검증 파이프라인 구동
# 내부에서 지정된 에포크(5회)만큼 train_step()과 test_step()이 맞물려 돌며 실시간 로그를 출력함
engine.train(model=model,
             train_dataloader=train_dataloader,
             test_dataloader=test_dataloader,
             loss_fn=loss_fn,
             optimizer=optimizer,
             epochs=NUM_EPOCHS,
             device=device)

# 10. 타이머 종료 및 최종 훈련 소요 시간 계산/출력
end_time = timer()
print(f"[INFO] Total training time: {end_time-start_time:.3f} seconds")

# 11. utils 모듈의 save_model() 함수를 사용해 학습된 최적 모델의 가중치 상태(state_dict)를 파일로 영구 저장
utils.save_model(model=model,
                 target_dir="models",
                 model_name="05_going_modular_script_mode_tinyvgg_model.pth")

Writing going_modular/train.py


In [ ]:
!python going_modular/train.py

  0% 0/5 [00:00<?, ?it/s]Epoch: 1 | train_loss: 1.0997 | train_acc: 0.2773 | test_loss: 1.1310 | test_acc: 0.2812
 20% 1/5 [00:01<00:04,  1.17s/it]Epoch: 2 | train_loss: 1.1104 | train_acc: 0.3164 | test_loss: 1.1291 | test_acc: 0.1979
 40% 2/5 [00:02<00:02,  1.02it/s]Epoch: 3 | train_loss: 1.1088 | train_acc: 0.2930 | test_loss: 1.0915 | test_acc: 0.2491
 60% 3/5 [00:02<00:01,  1.06it/s]Epoch: 4 | train_loss: 1.0922 | train_acc: 0.4141 | test_loss: 1.0912 | test_acc: 0.4432
 80% 4/5 [00:03<00:00,  1.10it/s]Epoch: 5 | train_loss: 1.0899 | train_acc: 0.4844 | test_loss: 1.0904 | test_acc: 0.3826
100% 5/5 [00:04<00:00,  1.08it/s]
[INFO] Total training time: 4.624 seconds
[INFO] Saving model to: models/05_going_modular_script_mode_tinyvgg_model.pth


We finish with a saved image classification model at `models/05_going_modular_cell_mode_tinyvgg_model.pth`.

See exercises and extra-curriculum in section 05 of the Learn PyTorch book - https://www.learnpytorch.io/05_pytorch_going_modular/#exercises

